lib/web_ui/dev/test_platform.dart

```diff
@@ -137,58 +142,78 @@ class BrowserPlatform extends PlatformPlugin {
     final Map<String, dynamic> requestData = json.decode(payload);
     final String filename = requestData['filename'];
     final bool write = requestData['write'];
-    final String result = await _diffScreenshot(filename, write);
+    final Map<String, dynamic> region = requestData['region'];
+    final String result = await _diffScreenshot(filename, write, region);
     return shelf.Response.ok(json.encode(result));
   }
 
-  Future<String> _diffScreenshot(String filename, bool write) async {
+  Future<String> _diffScreenshot(String filename, bool write, [ Map<String, dynamic> region ]) async {
     const String _kGoldensDirectory = 'test/golden_files';
+
+    // Bail out fast if golden doesn't exist, and user doesn't want to create it.
+    final File file = File(p.join(_kGoldensDirectory, filename));
+    if (!file.existsSync() && !write) {
+      return '''
+Golden file $filename does not exist.
+
+To automatically create this file call matchGoldenFile('$filename', write: true).
+''';
+    }
+
     final wip.ChromeConnection chromeConnection =
         wip.ChromeConnection('localhost', _kChromeDevtoolsPort);
     final wip.ChromeTab chromeTab = await chromeConnection.getTab(
         (wip.ChromeTab chromeTab) => chromeTab.url.contains('localhost'));
     final wip.WipConnection wipConnection = await chromeTab.connect();
+
+    Map<String, dynamic> captureScreenshotParameters = null;
+    if (region != null) {
+      captureScreenshotParameters = {
+        'format': 'png',
+        'clip': {
+          'x': region['x'],
```

Comment:
The clip is specified in pixels, and the test window should have a devicePixelRation of 1.0, right? Otherwise I think this would need some math

Reply:
Correct. This works because by default we have a devicePixelRatio of 1.

The clip is specified in [device independent pixels](https://chromedevtools.github.io/devtools-protocol/tot/Page#type-Viewport). The `scale` seems to let us change something *similar* to DPI (if I make scale: 2 I get screenshots that are 2x the size I requested), **however** reading docs it's not entirely clear to me that it'd trigger other things, like mediaqueries. I think this `scale` is similar to the browser "zoom" level, and we may always want it to be 1 (but I haven't verified this).

For us to emulate different DPIs, we probably need to set it with [Emulation.setDeviceMetricsOverride](https://chromedevtools.github.io/devtools-protocol/tot/Emulation#method-setDeviceMetricsOverride) (and friends) so mediaqueries and the rest of the browser behave as expected.